interface/ceed-qfunction.c

```diff
@@ -67,15 +79,83 @@ int CeedQFunctionCreateInterior(Ceed ceed, CeedInt vlength,
   (*qf)->refcount = 1;
   (*qf)->vlength = vlength;
   (*qf)->function = f;
-  ierr = CeedCalloc(strlen(focca)+1, &focca_copy); CeedChk(ierr);
-  strncpy(focca_copy, focca, strlen(focca)+1);
-  (*qf)->focca = focca_copy;
+  ierr = CeedCalloc(strlen(source)+1, &source_copy); CeedChk(ierr);
+  strncpy(source_copy, source, strlen(source)+1);
+  (*qf)->sourcepath = source_copy;
   ierr = CeedCalloc(16, &(*qf)->inputfields); CeedChk(ierr);
   ierr = CeedCalloc(16, &(*qf)->outputfields); CeedChk(ierr);
   ierr = ceed->QFunctionCreate(*qf); CeedChk(ierr);
   return 0;
 }
 
+/**
+  @brief Register a gallery QFunction
+
+  @param name   Name for this backend to respond to.
+  @param init   Initialization function called by CeedQFunctionInit() when the
+                    QFunction is selected.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedQFunctionRegister(const char *name, const char *source,
+                          CeedInt vlength, CeedQFunctionUser f,
+                          int (*init)(Ceed, const char *, CeedQFunction)) {
+  if (num_qfunctions >= sizeof(qfunctions) / sizeof(qfunctions[0])) {
+    return CeedError(NULL, 1, "Too many gallery QFunctions");
+  }
+  strncpy(qfunctions[num_qfunctions].name, name, strlen(name)+1);
```

Comment:
The length `n` is meant to be the length of the destination buffer (but `strncpy` is error-prone because it doesn't NULL terminate the result if there isn't enough space. You'll often see `strncpy(dest, src, dest_size); dest[dest_size-1] = 0;` to ensure a NULL-terminated result. BSDs provide `strlcpy` in `string.h`, but it isn't part of the standard. C11 offers `strcpy_s`, which is the interface you'd like to have, but it isn't standard.